### ISSUE TITLE
LDAP configuration - Invalid DN syntax

### ISSUE THREAD

**Issue by vfrans** [dpgaspar/Flask-AppBuilder#189]
Hello everyone,

I have a problem configuring FAB for LDAP.

The problem looks related to binding to LDAP.

#### 1. LDAP Setup
I've created a user in ldap called testfab here is the ldapsearch:
> ldapsearch -x -H ldaps://ldap.test.com -b "cn=testfab,ou=people,dc=test,dc=com" -D "cn=testfab,ou=people,dc=test,dc=com" -w\<password\>

##### Returns:
> dn: cn=testfab,ou=people,dc=test,dc=com
> mail: testfab@test.com
> userPassword:: \<hashed password\>
> displayName: Test Fab
> uid: testfab
> givenName: Test
> objectClass: inetOrgPerson
> sn: Fab
> cn: testfab

#####  config.py:
> AUTH_LDAP_SERVER = "ldap://ldap.test.com"
> AUTH_LDAP_SEARCH = "ou=people,dc=test,dc=com"
> AUTH_LDAP_UID_FIELD = "cn"
> AUTH_LDAP_FIRSTNAME_FIELD = "givenName"
> AUTH_LDAP_LASTNAME_FIELD = "sn"
> AUTH_LDAP_EMAIL_FIELD = "mail"

#### 3. Tests
After configuring ldap, entering a user/pass (login=testfab) present in ldap in the login box does:
##### slapd output:
> conn=45143 fd=21 ACCEPT from IP=XX.XX.XX.XX:52781 (IP=0.0.0.0:389)
> conn=45143 op=0 do_bind: invalid dn (testfab)
> conn=45143 op=0 RESULT tag=97 err=34 text=invalid DN
> conn=45143 op=1 UNBIND
> conn=45143 fd=21 closed

##### FAB debug output:
> 2015-07-15 09:37:05,888:DEBUG:flask_appbuilder.security.manager:LDAP bind with: testfab
> 2015-07-15 09:37:05,974:ERROR:flask_appbuilder.security.manager:LDAP Error Invalid DN syntax

#### 4. More tests!
To me it looks like FAB is trying to bind using as DN the text i typed in the login filed. I then tried to put my DN (cn=testfab,ou=people,dc=test,dc=com) as login to see what happens:
##### slapd output:
> conn=45155 fd=21 ACCEPT from IP=XX.XX.XX.XX:52808 (IP=0.0.0.0:389)
> conn=45155 op=0 BIND dn="cn=testfab,ou=people,dc=test,dc=com" method=128
> conn=45155 op=0 BIND dn="cn=testfab,ou=people,dc=test,dc=com" mech=SIMPLE ssf=0
> conn=45155 op=0 RESULT tag=97 err=0 text=
> conn=45155 op=1 UNBIND
> conn=45155 fd=21 closed

##### FAB debug output:
> 2015-07-15 09:46:14,959:DEBUG:flask_appbuilder.security.manager:LDAP bind with: cn=testfab,ou=people,dc=test,dc=com

No more invalid DN! But still FAB doesn't let me in and act like if the user/password was wrong probably because it's looking for a user called cn=testfab,ou=people,dc=test,dc=com when he should be looking for user called testfab (inside a CN field).
> Invalid login. Please try again.

I don't know if i am using FAB correctly, if not feel free to tell me :)

**Comment by dpgaspar** [comment]
Let run some tests on this. I'll give you an answer ASAP

**Comment by dpgaspar** [comment]
Are you using
   
 AUTH_USER_REGISTRATION = True 

on your config?

**Comment by dpgaspar** [comment]
Try this on your config:

``` python
AUTH_TYPE = AUTH_LDAP
AUTH_USER_REGISTRATION = False
AUTH_LDAP_SERVER = "ldap://ldap.test.com"
AUTH_LDAP_UID_FIELD = "uid"
```
Then login using the uid property of LDAP. on your example testfab (you must create first the testfab user on the DB)

**Comment by vfrans** [comment]
Thanks for the reply! Unfortunately, it doesn't work.

> INFO:flask_appbuilder.base:Registering class MyView on menu Method1
> [16/Jul/2015 09:37:25] "GET /login/ HTTP/1.1" 200 -
> DEBUG:flask_appbuilder.security.manager:LDAP bind with: testfab
> ERROR:flask_appbuilder.security.manager:LDAP Error Invalid DN syntax
> [16/Jul/2015 09:37:30] "POST /login/ HTTP/1.1" 302 -
> [16/Jul/2015 09:37:30] "GET /login/ HTTP/1.1" 200 -

With config

> AUTH_TYPE = AUTH_LDAP
> AUTH_USER_REGISTRATION = False
> AUTH_LDAP_SERVER = "ldap://ldap.test.com"
> AUTH_LDAP_SEARCH = "ou=people,dc=test,dc=com"
> AUTH_LDAP_UID_FIELD = "uid"
> AUTH_LDAP_FIRSTNAME_FIELD = "givenName"
> AUTH_LDAP_LASTNAME_FIELD = "sn"
> AUTH_LDAP_EMAIL_FIELD = "mail"

While db logging is possible and this is in the db:

> mysql> select id,username,active from ab_user;
> +----+----------+--------+
> | id | username | active |
> +----+----------+--------+
> |  1 | admin    |      1 |
> |  2 | testfab  |      1 |
> +----+----------+--------+

**Comment by dpgaspar** [comment]
mmmm, i'm not getting this error, but i'm using AD for LDAP server. What LDAP server are you using?

**Comment by vfrans** [comment]
I'm using OpenLDAP. To create the user on the database, i should just change the auth to DB, create the user, and change it back to LDAP, right?



—

**Comment by dpgaspar** [comment]
Yes, but you can just use fabmanager create-admin keeping the config as is.

**Comment by dpgaspar** [comment]
Still not been able to perform full tests. But i think i know whats the problem. I've allways runned tests using MS AD, but OpenLDAP need to bind with dn. I'll correct it and release a minor ASAP.

Thanks for reporting this.

**Comment by dpgaspar** [comment]
But first, could you please try the following:

on your config just put the following related with AUTH:

AUTH_TYPE = AUTH_LDAP
AUTH_LDAP_SERVER = "ldap://ldap.test.com"

Next create a user on the DB with username: testfab@test.com

Test login with testfab@test.com.

**Comment by vfrans** [comment]
Ok, i've tried it,

- Dropped the database and recreated it, config is as you requested.
- Created the user testfab@test.com using create-admin.
- Started FAB and tried to login using testfab@test.com as login.

Result:
slapd log
> conn=60495 fd=21 ACCEPT from IP=XX.XX.XX.XX:58645 (IP=0.0.0.0:389)
> conn=60495 op=0 do_bind: invalid dn (testfab@test.com)

fab log:
2015-07-20 09:35:18,412:DEBUG:flask_appbuilder.security.manager:LDAP bind with: testfab@test.com
2015-07-20 09:35:18,893:ERROR:flask_appbuilder.security.manager:LDAP Error Invalid DN syntax

**Comment by dpgaspar** [comment]
Right, thanks. I think i have nailed the problem, but still have not tested it with OpenLDAP.

The problem is that AD accepts **ldap.bind** with usernames, and RFC works only with DN. So i've added two more config key for LDAP and you must add these, first identify a user for whom all queries will run:

``` python
AUTH_LDAP_BIND_USER = "CN=binuser,OU=Familia,dc=casa,dc=local"
AUTH_LDAP_BIND_PASSWORD = "password1"
```

So when these keys exist, the auth method will first bind with BIND_USER and BIND_PASSWORD, next will search for your user using AUTH_LDAP_UID, and use the DN from the search to bind again with the user.

Could you please clone branch 1.4.4, install the 1.4.4 version, and run some tests with your example?

You can also download and unzip 1.4.4, after activating your virtualenv:

``` bash
$ wget https://github.com/dpgaspar/Flask-AppBuilder/archive/1.4.4.zip
$ unzip 1.4.4.zip
$ cd Flask-AppBuilder-1.4.4
$ python setup.py install
```

**Comment by vfrans** [comment]
Getting closer but still not :)

In fact OpenLDAP also require a search base, not just an UID field to look for.

After creating the user, configuring as asked it turns out this way:

FAB output:
> DEBUG:flask_appbuilder.security.manager:LDAP indirect bind with: CN=testfab,OU=people,dc=test,dc=com
> DEBUG:flask_appbuilder.security.manager:LDAP BIND indirect OK
> ERROR:flask_appbuilder.security.manager:LDAP Error No such object

And slapd's log:

> conn=62040 fd=24 ACCEPT from IP=XX.XX.XX.XX:63351 (IP=0.0.0.0:389)
> conn=62040 op=0 BIND dn="cn=testfab,ou=people,dc=test,dc=com" method=128
> conn=62040 op=0 BIND dn="cn=testfab,ou=people,dc=test,dc=com" mech=SIMPLE ssf=0
> conn=62040 op=0 RESULT tag=97 err=0 text=
> conn=62040 op=1 SRCH base="" scope=2 deref=0 filter="(uid=testfab)"
> conn=62040 op=1 SRCH attr=givenName sn mail
> conn=62040 op=1 SEARCH RESULT tag=101 err=32 nentries=0 text=
> conn=62040 op=2 UNBIND
> conn=62040 fd=24 closed

See the SRCH base="" ?

It should contain "ou=people,dc=test,dc=com" that way you'd get all the records in this OU including the user testfab.

Maybe you could add an option to configure the search base?

Thanks a lot for your help!

**Comment by dpgaspar** [comment]
Thanks for the output, tests and patience, could you please use also:

AUTH_LDAP_SEARCH = "dc=test,dc=com"

**Comment by vfrans** [comment]
It works :)

Great work, many thanks!

**Comment by dpgaspar** [comment]
Included on 1.4.4 and version released.
